mtags/src/main/scala/scala/meta/internal/pc/OverrideCompletions.scala

```diff
@@ -0,0 +1,506 @@
+        else if (sym.isStable) "val "
+        else "def "
+
+      val asciOverrideDef: String = {
+        if (sym.isAbstract) keyword
+        else s"${overrideKeyword}${keyword}"
+      }
+
+      val overrideDef: String = metalsConfig.overrideDefFormat() match {
+        case OverrideDefFormat.Unicode =>
+          if (sym.isAbstract) "🔼 "
+          else "⏫ "
+        case _ => asciOverrideDef
+      }
+
+      val name: String = Identifier(sym.name)
+
+      val filterText: String = s"${overrideKeyword}${lzy}${keyword}${name}"
+
+      // if we had no val or def then filter will be empty
+      def toMember = new OverrideDefMember(
+        label,
+        edit,
+        filterText,
+        sym,
+        history.autoImports(
+          pos,
+          importContext,
+          autoImport.offset,
+          autoImport.indent,
+          autoImport.padTop
+        ),
+        details
+      )
+
+      private def label = overrideDef + name + signature
+      private def details = asciOverrideDef + name + signature
+      private def signature = printer.defaultMethodSignature()
+      private def edit = new l.TextEdit(
+        range,
+        if (clientSupportsSnippets && shouldMoveCursor) {
+          s"$filterText$signature = $${0:???}"
+        } else {
+          s"$filterText$signature = ???"
+        }
+      )
+    }
+
+    typed.tpe.members.iterator.toList
+      .filter(isOverridableMethod)
+      .map(OverrideCandidate.apply)
+      .map(_.toMember)
+  }
+
+  private def toEdits(
+      allAbstractMembers: List[OverrideDefMember]
+  ): (List[l.TextEdit], Set[l.TextEdit]) = {
+    allAbstractMembers.foldLeft(
+      (List.empty[l.TextEdit], Set.empty[l.TextEdit])
+    ) { (editsAndImports, overrideDefMember) =>
+      val edits = overrideDefMember.edit :: editsAndImports._1
+      val imports = overrideDefMember.autoImports.toSet ++ editsAndImports
+        ._2
+      (edits, imports)
+    }
+  }
+
+  // NOTE(gabro): sym.isVar does not work consistently across Scala versions
+  // Specifically, it behaves differently between 2.11 and 2.12/2.13
+  // This check is borrowed from
+  // https://github.com/scala/scala/blob/f389823ef0416612a0058a80c1fe85948ff5fc0a/src/reflect/scala/reflect/internal/Symbols.scala#L2645
+  private def isVarSetter(sym: Symbol): Boolean =
+    !sym.isStable && !sym.isLazy && sym.isAccessor
+
+  def implementAllAt(pos: Position, text: String): List[l.TextEdit] = {
+    // make sure the compilation unit is loaded
+    typedTreeAt(pos)
+
+    lastVisistedParentTrees match {
+
+      // class Foo extends Bar {}
+      // ~~~~~~~~~~~~~~~~~~~~~~~~
+      case (c: ClassDef) :: _ =>
+        val t = c.impl
+        val typed = typedTreeAt(t.pos)
+        implementAll(
+          typed,
+          inferEditPosition(text, t).toLSP,
+          t,
+          text,
+          true,
+          _ => true
+        )
+
+      // object Foo extends Bar {}
+      // ~~~~~~~~~~~~~~~~~~~~~~~~
+      case (m: ModuleDef) :: _ =>
```

Comment:
all the cases are nearly identical in their implementation, the only difference is how they extract a tree. Can we maybe extract it to a utility method taking a Tree as input?

Reply:
done in https://github.com/scalameta/metals/pull/1379/commits/761a52952d7d9138cba683bbbd1c638723889903
Thank you for your advice, it's much clear :)